### PR TITLE
Fix for incorrect computation of batched alignment in transducers

### PR DESCRIPTION
**Pull request by Kipok**
# What does this PR do ?

The original code had a bug because of which whenever at least one batch element predicted non-blank symbol, all blank elements would duplicate the last prediction in their `.alignments` variable.

Additionally, I've unified `_greedy_decode_blank_as_pad` and `_greedy_decode_masked` into a single function with an if check. The two methods had 99% of the same code except 5 lines or so, so imo it does not make sense to have them as fully separate code paths. Please let me know if for some reason we don't want to have this change.

**Collection**: ASR

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py [hunk 1]
```diff
@@ -811,211 +819,6 @@ def _greedy_decode_blank_as_pad(
 
         return hypotheses
 
-    def _greedy_decode_masked(
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:814]: Why is this being removed?

File: nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py [hunk 2]
```diff
@@ -679,8 +671,21 @@ def _greedy_decode_blank_as_pad(
                     if time_idx == 0 and symbols_added == 0 and hidden is None:
                         g, hidden_prime = self._pred_step(self._SOS, hidden, batch_size=batchsize)
                     else:
-                        # Perform batch step prediction of decoder, getting new states and scores ("g")
-                        g, hidden_prime = self._pred_step(last_label, hidden, batch_size=batchsize)
+                        if self.decoder.blank_as_pad:
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:674]: Please revert, we dont want to merge these two functions.

File: nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py [hunk 3]
```diff
@@ -650,7 +642,8 @@ def _greedy_decode_blank_as_pad(
             # Last Label buffer + Last Label without blank buffer
             # batch level equivalent of the last_label
             last_label = torch.full([batchsize, 1], fill_value=self._blank_index, dtype=torch.long, device=device)
-
+            if not self.decoder.blank_as_pad:
+                last_label_without_blank = last_label.clone()
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:646]: Add else case setting var to None

**Review by titu1994 — CHANGES_REQUESTED** [review]
Revert merge of the two functions please.

File: nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py (hunk 2 above)
Comment (Kipok) [reply nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:674]: Can you please explain why we don't want to merge them? They are different in just a few lines, which can be handled with an if-else block (or a new function if you'd prefer that way). And I had to apply the same fix to both functions, so I thought it makes sense to unify them. If we unify them and in the future we need to make some changes, it's going to be in a single place and not in two places, which is easy to forget.

Comment (titu1994) [reply nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py:674]: They share similar logic but not concept. Normally the prednet is not supposed to have the blank token in it's vocab. We abuse the notation by using the pytorch embeddings mechanism for masking one token in NeMo for efficiency sake but the academically correct code is the second one. 

They should not be merged.

Comment (Kipok) [reply nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py]: Rolled-back

**Comment by Kipok** [comment]
I think the same changes would need to go for multi-blank RNNT as well, but it's better if @hainan-xv can confirm. It should be easy to test using the same test code I added, assuming there is some pretrained model we can use

**Review by titu1994 — APPROVED** [review]
Looks great, thanks ! @hainan-xv please look into the multi blank rnnt version and send a PR
